core/dbt/graph/selector.py

```diff
@@ -162,12 +187,53 @@ def filter_selection(self, selected: Set[UniqueId]) -> Set[UniqueId]:
             unique_id for unique_id in selected if self._is_match(unique_id)
         }
 
-    def expand_selection(self, selected: Set[UniqueId]) -> Set[UniqueId]:
-        """Perform selector-specific expansion."""
+    def expand_selection(
+        self, selected: Set[UniqueId], greedy: bool = False
+    ) -> Tuple[Set[UniqueId], Set[UniqueId]]:
+        # Test selection can expand to include an implicitly/indirectly selected test.
+        # In this way, `dbt test -m model_a` also includes tests that directly depend on `model_a`.
+        # Expansion has two modes, GREEDY and NOT GREEDY.
+        #
+        # GREEDY mode: If ANY parent is selected, select the test. We use this for EXCLUSION.
+        #
+        # NOT GREEDY mode:
+        #  - If ALL parents are selected, select the test.
+        #  - If ANY parent is missing, return it separately. We'll keep it around
+        #    for later and see if its other parents show up.
+        # We use this for INCLUSION.
+
+        direct_nodes = selected
```

Comment:
I think that using `selected` directly here might not be a _great_ idea because `direct_nodes` is just a pointer to the `selected` function argument. On line 214 below, we're calling the `direct_nodes.add` function which is going to mutate the `selected` variable and could possibly cause problems in any function calls up the call stack.

This would _not_ cause any problems today since the caller of this method looks like:
```
        direct_nodes, indirect_nodes = self.expand_selection(
            selected=(collected | neighbors),
            greedy=spec.greedy
        )
```

so, the `selected` argument is an expression and is not assigned to a variable. Imagine if this logic changed in the future to look more like:

```
        to_select = (collected | neighbors)
        direct_nodes, indirect_nodes = self.expand_selection(
            selected=to_select,
            greedy=spec.greedy
        )
        some_function(to_select)
```

If this happened, then the value of `to_select` would be changed after the call to `expand_selection`! That kind of behavior would probably not be expected, and it would probably be a tough bug to track down!

Good news is that it's a quick change, I think you can make a copy of the set by changing this line to:
```
direct_nodes = set(selected)
```

That way, this method won't have any "side effects"

Reply:
You're 100% right! Including `direct_nodes = selected` was my attempt at not mutating the original argument value, but that's obviously not what it does. Wrapping with `set()` makes a ton of sense.